alpha/ui.cpp

```diff
@@ -31,69 +31,157 @@ void _TS_UI::begin()
+        TS_HAL.lcd_sleep(false);
+        TS_HAL.lcd_brightness(brightness);
+      }
 
-  while(true)
-  {
-    // don't update contents if lcd is not turned on
-    if(!SavePower)
-    {
       TS_DateTime datetime;
       TS_HAL.rtc_get(datetime);
+
+      TS_HAL.lcd_cursor(0, 0);
+      // TODO: Does not work with F()
+
+      TS_HAL.lcd_printf("%04d-%02d-%02d ", datetime.year, datetime.month, datetime.day);
+      TS_HAL.lcd_printf("%02d:%02d:%02d\n", datetime.hour, datetime.minute, datetime.second);
+      TS_HAL.lcd_printf("Battery: %d%%", TS_HAL.power_get_batt_level());
+      if (TS_HAL.power_is_charging()) {
+        TS_HAL.lcd_printf(", Charging");
+      }
+      TS_HAL.lcd_printf("\n");
       
-      TS_HAL.lcd_cursor(0, 15);
-      // TODO: does not work with F()
-      TS_HAL.lcd_printf("Date: %04d-%02d-%02d\n",     datetime.year, datetime.month, datetime.day);
-      TS_HAL.lcd_printf("Time: %02d : %02d : %02d\n", datetime.hour, datetime.minute, datetime.second);
-      TS_HAL.lcd_printf("Battery: %d%%  ", TS_HAL.power_get_batt_level(), NULL, NULL);
-      if (TS_HAL.power_is_charging()) 
-      {
-        TS_HAL.lcd_printf("Status: Charging    ");
-      } else {
-        TS_HAL.lcd_printf("Status: Not Charging");
+      for (int i = 0; i < 3; i++) {
+        if (selected == i) {
+          TS_HAL.lcd_printf(" [%s]  \n", options[i]);
+        } else if (cursor == i) {
+          TS_HAL.lcd_printf("> %s   \n", options[i]);
+        } else {
+          TS_HAL.lcd_printf("  %s   \n", options[i]);
+        }
       }
-    }
 
-    // stackHighWaterMark = uxTaskGetStackHighWaterMark(NULL);
-    // Serial.print("UI: ");
-    // Serial.println(stackHighWaterMark);
+      stackHighWaterMark = uxTaskGetStackHighWaterMark(NULL);
+      Serial.print("UI: ");
+      Serial.println(stackHighWaterMark);
+
+      // Select option 3 to sleep.
+      if (selected == 2) {
+        TS_HAL.lcd_sleep(true);
+        savePowerStart = millis();
+      }
+    }
 
-    if (TS_HAL.btn_a_get() == TS_ButtonState::Short) 
-    {
+    if (TS_HAL.btn_a_get() == TS_ButtonState::Short) {
       Serial.println("Button A pressed");
     }
 
-    if (TS_HAL.btn_b_get() == TS_ButtonState::Short) 
-    {
+    if (TS_HAL.btn_b_get() == TS_ButtonState::Short) {
       Serial.println("Button B pressed");
     }
 
     TS_ButtonState powerButtonState = TS_HAL.btn_power_get();
-    if (powerButtonState == TS_ButtonState::Short) 
-    {
+    if (powerButtonState == TS_ButtonState::Short) {
       Serial.println("Power button short press");
-    } else if (powerButtonState == TS_ButtonState::Long) 
-    {
+      TS_HAL.lcd_cursor(0, 0);
+      TS_HAL.lcd_printf("Power short");
+    } else if (powerButtonState == TS_ButtonState::Long) {
       Serial.println("Power button long press");
+      TS_HAL.lcd_cursor(0, 0);
+      TS_HAL.lcd_printf("Power long");
     }
 
-    TS_HAL.sleep(TS_SleepMode::Task, 250);
+    TS_HAL.sleep(TS_SleepMode::Task, 20);
```

Comment:
50fps? :o

Reply:
Improvement: should sleep take into account render time? e.g. if render interval is 20ms, shouldn't it count how many ms has elapsed since last render?

Reply:
there is no render if there is no user activity. this is button check interval. so yes, if users can press the button at 50 hertz, we will render at 50 hertz.